File: src/chipschallenge/blockreactions/IceTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Inventory.Boots;
import chipschallenge.Move.Moves;

public class IceTo extends BlockReaction {

    private IceTo() {
    }
    private static IceTo mInstance = null;

    public static synchronized IceTo getInstance() {
        if (mInstance == null) {
            mInstance = new IceTo();
        }
        return mInstance;
    }

    @Override
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        // Nothing
    }

    @Override
    public boolean canMove(Block moving, Block standing) {
        return true;
    }

    @Override
    public Moves causesSlip(Block moving, Block standing) {
        if(moving.isChip() && hasBoots(Boots.ICESKATES))
            return null;
        else
            return moving.getFacing();
    }
}

File: ChipsChallenge.Shared/Blockreactions/IceTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Boots = Inventory.Boots;
    using Moves = Move.Moves;

    public class IceTo : BlockReaction
    {
        private static IceTo instance;

        private IceTo()
        {
        }

        public static IceTo Instance
        {
            get
            {
                lock (typeof(IceTo))
                {
                    return instance ?? (instance = new IceTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }

        public override Moves? CausesSlip(Block moving, Block standing)
        {
            if (moving.Chip && HasBoots(Boots.ICESKATES))
            {
                return null;
            }

            return moving.Facing;
        }
    }
}